File: Source/Meowtrix.PixivApi/Authentication/PixivAuthentication.cs
using System;
using System.Buffers.Text;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using IdentityModel.Client;
using Meowtrix.PixivApi.Json;

namespace Meowtrix.PixivApi.Authentication;

/// <summary>
/// Provide OAuth-based authentication for Pixiv.
/// </summary>
public static class PixivAuthentication
{
    private const string ClientId = "MOBrBDS8blbauoSck0ZfDbtuzpyT";
    private const string ClientSecret = "lsACyCD94FhDUtGTXi3QzcFE2uU1hqtDaKeqrdwj";
    private const string UserAgent = "PixivAndroidApp/5.0.166 (Android 12.0)";
    private const string AuthUrl = "https://oauth.secure.pixiv.net/auth/token";

#if NET
    private static ReadOnlySpan<byte> HashSecret => "28c1fdd170a5204386cb1313c7077b34f83e4aaf4aa829ce78c231e05b0bae2c"u8;
#else
    private const string HashSecret = "28c1fdd170a5204386cb1313c7077b34f83e4aaf4aa829ce78c231e05b0bae2c";
#endif

    private static (string Time, string ClientHash) BuildClientTimeHeader(DateTimeOffset timestamp)
    {
        string requestTime = timestamp.ToString(@"yyyy-MM-dd\THH\:mm\:ssK");
#if NET
        Span<byte> clientHash = stackalloc byte[MD5.HashSizeInBytes];
        MD5.HashData([.. Encoding.UTF8.GetBytes(requestTime), .. HashSecret], clientHash);
#else
        using var md5 = MD5.Create();
        byte[] clientHash = md5.ComputeHash(Encoding.UTF8.GetBytes(requestTime + HashSecret));
#endif

#if NET9_0_OR_GREATER
        return (requestTime, Convert.ToHexStringLower(clientHash));
#elif NET
        return (requestTime, Convert.ToHexString(clientHash).ToLowerInvariant());
#else
        return (requestTime, BitConverter.ToString(clientHash).Replace("-", "").ToLowerInvariant());
#endif
    }

    /// <summary>
    /// Get authentication infomation with refresh token.
    /// </summary>
    /// <param name="httpMessageInvoker">The http client to send authentication request.</param>
    /// <param name="refreshToken">The refresh token.</param>
    /// <param name="timeProvider">The time provider to verify client time.</param>
    /// <param name="cancellationToken">The cancellation token.</param>
    /// <returns>The authentication result.</returns>
    /// <exception cref="AuthenticationFailedException">When authentication failed.</exception>
    public static async Task<PixivAuthenticationResult> AuthWithRefreshTokenAsync(
        HttpMessageInvoker httpMessageInvoker,
        string refreshToken,
        TimeProvider? timeProvider = null,
        CancellationToken cancellationToken = default)
    {
        timeProvider ??= TimeProvider.System;
        (string requestTime, string clientHash) = BuildClientTimeHeader(timeProvider.GetUtcNow());

        using var request = new RefreshTokenRequest
        {
            Address = AuthUrl,
            ClientId = ClientId,
            ClientSecret = ClientSecret,
            RefreshToken = refreshToken,
            Headers =
            {
                { "User-Agent", UserAgent },
                { "X-Client-Time", requestTime },
                { "X-Client-Hash", clientHash },
            },
#if NET
            VersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,
#endif
        };

        var response = await httpMessageInvoker.RequestRefreshTokenAsync(request, cancellationToken).ConfigureAwait(false);
        return response.CheckAuthenticationResult(timeProvider);
    }

    /// <summary>
    /// Prepare the parameters for doing browser-based login.
    /// </summary>
    /// <returns>The parameters for login.</returns>
    /// <remarks>
    /// To complete the login process, navigate to <c>LoginUrl</c> in browser, listen to <c>pixiv://....?code=....</c> ,
    /// get authentication code from the <c>code</c> query parameter, and invoke <see cref="CompleteWebLoginAsync"/>.
    /// </remarks>
    public static (string CodeVerify, string LoginUrl) PrepareWebLogin()
    {
#if NET
        Span<byte> bytes = stackalloc byte[36];
        RandomNumberGenerator.Fill(bytes);

        string codeVerifyString = Convert.ToBase64String(bytes);

        Span<byte> codeVerify = stackalloc byte[48];
        Encoding.UTF8.GetBytes(codeVerifyString, codeVerify);

        Span<byte> sha = stackalloc byte[SHA256.HashSizeInBytes];
        SHA256.HashData(codeVerify, sha);
#else
        using var rng = RandomNumberGenerator.Create();
        byte[] bytes = new byte[36];
        rng.GetBytes(bytes);

        string codeVerifyString = Convert.ToBase64String(bytes);
        byte[] codeVerify = Encoding.UTF8.GetBytes(codeVerifyString);

        using var sha256 = SHA256.Create();
        byte[] sha = sha256.ComputeHash(codeVerify);

#endif

#if NET9_0_OR_GREATER
        Span<char> urlSafeCodeChallenge = stackalloc char[43];
        Base64Url.EncodeToChars(sha, urlSafeCodeChallenge);
#else
        string urlSafeCodeChallenge = Convert.ToBase64String(sha)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
#endif

        string loginUrl = $"https://app-api.pixiv.net/web/v1/login?code_challenge={urlSafeCodeChallenge}&code_challenge_method=S256&client=pixiv-android";

        return (codeVerifyString, loginUrl);
    }

    /// <summary>
    /// Complete browser-based authentication.
    /// </summary>
    /// <param name="httpMessageInvoker">The http client to send authentication request.</param>
    /// <param name="authorizationCode">The authentication code acquired from browser.</param>
    /// <param name="codeVerify">The verification code returned from <see cref="PrepareWebLogin"/>.</param>
    /// <param name="timeProvider">The time provider to verify client time.</param>
    /// <param name="cancellationToken">The cancellation token.</param>
    /// <returns>The authentication result.</returns>
    /// <exception cref="AuthenticationFailedException">When authentication failed.</exception>
    public static async Task<PixivAuthenticationResult> CompleteWebLoginAsync(
        HttpMessageInvoker httpMessageInvoker,
        string authorizationCode,
        string codeVerify,
        TimeProvider? timeProvider = null,
        CancellationToken cancellationToken = default)
    {
        timeProvider ??= TimeProvider.System;
        (string requestTime, string clientHash) = BuildClientTimeHeader(timeProvider.GetUtcNow());

        using var request = new AuthorizationCodeTokenRequest
        {
            Address = AuthUrl,
            ClientId = ClientId,
            ClientSecret = ClientSecret,
            Code = authorizationCode,
            RedirectUri = "https://app-api.pixiv.net/web/v1/users/auth/pixiv/callback",
            CodeVerifier = codeVerify,
            Headers =
            {
                { "User-Agent", UserAgent },
                { "X-Client-Time", requestTime },
                { "X-Client-Hash", clientHash },
            },
#if NET
            VersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,
#endif
        };

        var response = await httpMessageInvoker.RequestAuthorizationCodeTokenAsync(request, cancellationToken).ConfigureAwait(false);
        return response.CheckAuthenticationResult(timeProvider);
    }

    /// <summary>
    /// Gets authentication result from <see cref="IdentityModel"/> response.
    /// </summary>
    /// <param name="response">A response message from <see cref="IdentityModel"/>.</param>
    /// <param name="timeProvider">The time provide to calculate <see cref="PixivAuthenticationResult.ValidUntil"/>.</param>
    /// <returns>The authentication result.</returns>
    /// <exception cref="AuthenticationFailedException">When authentication failed.</exception>
    public static PixivAuthenticationResult CheckAuthenticationResult(this TokenResponse response, TimeProvider? timeProvider = null)
    {
        if (response.IsError)
        {
            throw new AuthenticationFailedException(response.Error, response.Exception);
        }

        if (response.AccessToken is null || response.RefreshToken is null)
        {
            throw new AuthenticationFailedException("No access token is provided.");
        }

        var validUntil = (timeProvider ?? TimeProvider.System).GetUtcNow().AddSeconds(response.ExpiresIn);
        var userInfo = response.Json?.GetProperty("user"u8).Deserialize(PixivJsonContext.Default.AuthUser)
            ?? throw new AuthenticationFailedException("No user info is provided.");
        return new(response.AccessToken, response.RefreshToken, validUntil, userInfo);
    }
}

public class AuthenticationFailedException(string? message, Exception? innerException = null) : Exception(message, innerException)
{
}

public record class PixivAuthenticationResult(string AccessToken, string RefreshToken, DateTimeOffset ValidUntil, AuthUser UserInfo);

File: Source/Meowtrix.PixivApiFS/PixivAuthentication.fs
namespace Meowtrix.PixivApiFS

open System
open System.Buffers.Text
open System.Net.Http
open System.Security.Cryptography
open System.Text
open System.Text.Json
open System.Threading.Tasks
open IdentityModel.Client

type PixivAuthenticationResult =
    { accessToken: string;
      refreshToken: string; 
      validUntil: DateTimeOffset;
      userInfo: AuthUser }

type AuthenticationFailedException(message: string | null, innerException: exn | null) =
    inherit Exception(message, innerException)

module PixivAuthentication =
    [<Literal>]
    let private ClientId = "MOBrBDS8blbauoSck0ZfDbtuzpyT"
    [<Literal>]
    let private ClientSecret = "lsACyCD94FhDUtGTXi3QzcFE2uU1hqtDaKeqrdwj"
    [<Literal>]
    let private UserAgent = "PixivAndroidApp/5.0.166 (Android 12.0)"
    [<Literal>]
    let private HashSecret = "28c1fdd170a5204386cb1313c7077b34f83e4aaf4aa829ce78c231e05b0bae2c"
    [<Literal>]
    let private AuthUrl = "https://oauth.secure.pixiv.net/auth/token"

    let private BuildClientTimeHeader (timestamp: DateTimeOffset) =
        let requestTime = timestamp.ToString @"yyyy-MM-dd\THH\:mm\:ssK"
        let clientHash =
            requestTime + HashSecret
            |> Encoding.UTF8.GetBytes
            |> Convert.ToHexStringLower
        (requestTime, clientHash)
        
    let CheckAuthenticationResult (response: TokenResponse) (timeProvider: TimeProvider) =
        if response.IsError
        then AuthenticationFailedException(response.Error, response.Exception) |> raise
        else
            match response.AccessToken, response.RefreshToken with
            | (null, _) | (_, null) -> raise(AuthenticationFailedException("No access token is provided.", null))
            | (accessToken, refreshToken) ->
                let validUntil = timeProvider.GetUtcNow().AddSeconds(response.ExpiresIn)
                let userInfo =
                    match response.Json with
                    | NonNullV jsonElement -> jsonElement.GetProperty("user") |> JsonSerializer.Deserialize<AuthUser>
                    | _ -> raise(AuthenticationFailedException("No user info is provided.", null))
                { accessToken = accessToken; refreshToken = refreshToken; validUntil = validUntil; userInfo = nonNull userInfo }
                
    let AuthWithRefreshTokenAsync (httpMessageInvoker: HttpMessageInvoker) refreshToken (timeProvider: TimeProvider) =
        async {
            let (requestTime, clientHash) = timeProvider.GetUtcNow() |> BuildClientTimeHeader
            use request = new RefreshTokenRequest(
                Address = AuthUrl,
                ClientId = ClientId,
                ClientSecret = ClientSecret,
                RefreshToken = refreshToken
            )
            request.Headers.Add("UserAgent",UserAgent)
            request.Headers.Add("X-Client-Time", requestTime)
            request.Headers.Add("X-Client-Hash", clientHash)
            let! response =
                httpMessageInvoker.RequestRefreshTokenAsync(request, Async.DefaultCancellationToken)
                |> Async.AwaitTask

            return CheckAuthenticationResult response timeProvider
        }

    let PrepareWebLogin() =
        let codeVerify =
            RandomNumberGenerator.GetBytes 36
            |> Convert.ToBase64String
        let codeChallenge =
            codeVerify
            |> Encoding.UTF8.GetBytes
            |> SHA256.HashData
            |> Base64Url.EncodeToString
        let loginUrl = $"https://app-api.pixiv.net/web/v1/login?code_challenge={codeChallenge}&code_challenge_method=S256&client=pixiv-android"
        (codeVerify, loginUrl)

    let CompleteWebLoginAsync (httpMessageInvoker: HttpMessageInvoker) authroizationCode codeVerify (timeProvider: TimeProvider) =
        async {
            let (requestTime, clientHash) = timeProvider.GetUtcNow() |> BuildClientTimeHeader
            use request = new AuthorizationCodeTokenRequest(
                Address = AuthUrl,
                ClientId = ClientId,
                ClientSecret = ClientSecret,
                Code = authroizationCode,
                RedirectUri = "https://app-api.pixiv.net/web/v1/users/auth/pixiv/callback",
                CodeVerifier = codeVerify
            )
            request.Headers.Add("UserAgent",UserAgent)
            request.Headers.Add("X-Client-Time", requestTime)
            request.Headers.Add("X-Client-Hash", clientHash)
            let! response =
                httpMessageInvoker.RequestAuthorizationCodeTokenAsync(request, Async.DefaultCancellationToken)
                |> Async.AwaitTask

            return CheckAuthenticationResult response timeProvider
        }

    type AccessTokenManager(initialAuthResult: PixivAuthenticationResult, ?timeProvider: TimeProvider) =
        let timeProvider = defaultArg timeProvider TimeProvider.System

        let mutable authResult = initialAuthResult
        member val RefreshThreshold = TimeSpan.FromMinutes(10L) with get, set

        member this.GetAccessTokenAsync(handler: HttpMessageHandler) =
            async {
                let authResultSnap = authResult
                if authResultSnap.validUntil - timeProvider.GetUtcNow() > this.RefreshThreshold
                then return authResultSnap.accessToken
                else
                    let httpMessageInvoker = new HttpMessageInvoker(handler, disposeHandler=false)
                    let! newAuthResult = AuthWithRefreshTokenAsync httpMessageInvoker authResultSnap.refreshToken timeProvider
                    authResult <- newAuthResult
                    return newAuthResult.accessToken
            }
